compiler/rustc_hir_analysis/src/hir_ty_lowering/mod.rs

```diff
@@ -2193,46 +2195,76 @@ impl<'tcx> dyn HirTyLowerer<'tcx> + '_ {
             && let hir::FnRetTy::Return(ty) = decl.output
             && let hir::TyKind::InferDelegation(_, _) = ty.kind
         {
-            try_emit("recursive delegation");
+            try_emit("recursive delegation is not supported yet");
         }
 
         let sig = self.tcx().fn_sig(sig_id).instantiate_identity();
         if sig.output().has_opaque_types() {
-            try_emit("delegation to a function with opaque type");
+            try_emit("delegation to a function with opaque type is not supported yet");
         }
 
         let sig_generics = self.tcx().generics_of(sig_id);
         let parent = self.tcx().parent(self.item_def_id());
         let parent_generics = self.tcx().generics_of(parent);
 
-        let parent_is_trait = (self.tcx().def_kind(parent) == DefKind::Trait) as usize;
+        let parent_def_kind = self.tcx().def_kind(parent);
+        let parent_is_trait = (parent_def_kind == DefKind::Trait) as usize;
         let sig_has_self = sig_generics.has_self as usize;
 
         if sig_generics.count() > sig_has_self || parent_generics.count() > parent_is_trait {
-            try_emit("delegation with early bound generics");
+            try_emit("delegation with early bound generics is not supported yet");
+        }
+
+        // There is no way to instantiate `Self` param for caller if
+        // 1. callee is a trait method && callee contains `Self` param
+        // 2. parent item isn't a trait or impl
+        if let DefKind::AssocFn = self.tcx().def_kind(sig_id)
+            && !matches!(parent_def_kind, DefKind::Impl { .. } | DefKind::Trait)
+            && self.tcx().associated_item(sig_id).container
+                == ty::AssocItemContainer::TraitContainer
+        {
+            struct HasSelfParamVisitor;
+
+            impl<'tcx> ty::TypeVisitor<TyCtxt<'tcx>> for HasSelfParamVisitor {
+                type Result = ControlFlow<()>;
+
+                fn visit_ty(&mut self, ty: Ty<'tcx>) -> Self::Result {
+                    // All early bound generics except `Self` are previously banned.
+                    // It is enough to check only the `ty::Param`.
+                    if matches!(ty.kind(), ty::Param(..)) {
+                        return ControlFlow::Break(());
+                    }
+
+                    ty.super_visit_with(self)
+                }
+            }
+
+            if sig.visit_with(&mut HasSelfParamVisitor).is_break() {
+                try_emit("Couldn't instantiate `Self` callee type");
```

Comment:
This message is oriented towards compiler writers instead of users.
(Also, diagnostic messages conventionally start with a lowercase letter.)